peer_gater.go

```diff
@@ -0,0 +1,252 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var PeerGaterRetainStats = 6 * time.Hour
+var PeerGaterQuiet = time.Minute
+
+type peerGater struct {
+	sync.Mutex
+
+	threshold, decay   float64
+	validate, throttle float64
+
+	lastThrottle time.Time
+
+	stats map[peer.ID]*peerGaterStats
```

Comment:
This looks like a potential OOM vector, with a 6 hour timeout.

Reply:
Also, maybe we want to taint the IP as well? It's pretty easy to roll a new peer ID.

Reply:
For filecoin, we may _also_ want to be able to use an arbitrary "keyer" function (e.g., a function that takes the peer ID and returns a miner ID).

Reply:
tainting the IP is a really good idea I think; I'll add it.

Re: OOM -- I don't know, the memory usage is very low, just a few bytes.

Reply:
so if we taint the IP and have a common stats object for all peers in the same IP, our memory usage per peer ID becomes just the peer ID itself and a pointer; attacker would have to generate billions of peer IDs in the 6 hour interval to make a dent in memory usage.

Reply:
OOM vector discussion: You'd need ~100M peer IDs to take ~4GiB of memory. That's over 4k new connections a second.

Reply:
I implemented per IP stat tracking in https://github.com/libp2p/go-libp2p-pubsub/pull/384/commits/cb4aae17443b622231f70a0c54ec3a6712fbce26.